File: cpp/0153.cpp
#include <vector>
#include <iostream>
using namespace std;


class Solution {
public:
    int findMin(vector<int>& nums) {
        int i = 0;
        int j = nums.size() - 1;
        while (i < j) {
            int k = (i + j) / 2;
            if (nums[j] > nums[k]) {
                j = k;
            } else {
                i = k+1;
            }
        }
        return nums[i];
    }
};

int main() {
    Solution sol;
    vector<int> nums;
    nums = {2, 1};
    cout << sol.findMin(nums) << endl;
    nums = {3, 4, 5, 1, 2};
    cout << sol.findMin(nums) << endl;
    nums = {4, 5, 6, 7, 0, 1, 2};
    cout << sol.findMin(nums) << endl;
    nums = {11, 13, 15, 17};
    cout << sol.findMin(nums) << endl;
}

File: python/0153.py
#  https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/


class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        left = 0
        right = len(nums) - 1

        while left < right:
            mid = (left + right) // 2
            if nums[mid] > nums[right]:
                left = mid + 1
            else:
                right = mid

        return nums[left]


sol = Solution()
assert sol.findMin([3, 4, 5, 1, 2]) == 1
assert sol.findMin([4,5,6,7,0,1,2]) == 0
assert sol.findMin([11,13,15,17]) == 11

